File: MainApp/src/main/java/com/sleticalboy/dailywork/http/CookieJarImpl.java
package com.sleticalboy.dailywork.http;

import android.content.Context;
import android.content.SharedPreferences;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import okhttp3.Cookie;
import okhttp3.CookieJar;
import okhttp3.HttpUrl;

/**
 * Created on 18-3-28.
 *
 * @author sleticalboy
 * @description Cookie 持久化「sp, db, file」
 */
public class CookieJarImpl implements CookieJar {

    private final SharedPreferences mPreferences;

    public CookieJarImpl(Context context) {
        if (context == null) {
            throw new NullPointerException("context == null");
        }
        mPreferences = context.getSharedPreferences("cookie-jar", Context.MODE_PRIVATE);
    }

    @Override
    public void saveFromResponse(HttpUrl url, List<Cookie> cookies) {
        final Set<String> values = new HashSet<>();
        for (final Cookie cookie : cookies) {
            values.add(cookie.toString());
        }
        mPreferences.edit().putStringSet(url.host(), values).apply();
    }

    @Override
    public List<Cookie> loadForRequest(HttpUrl url) {
        final Set<String> values = mPreferences.getStringSet(url.host(), new HashSet<>());
        List<Cookie> cookies = new ArrayList<>();
        for (final String value : values) {
            cookies.add(Cookie.parse(url, value));
        }
        return cookies;
    }
}

File: MainApp/src/main/java/com/sleticalboy/http/CookieJarImpl.kt
package com.binlee.http

import android.content.Context
import android.content.SharedPreferences
import okhttp3.Cookie
import okhttp3.CookieJar
import okhttp3.HttpUrl
import java.util.ArrayList
import java.util.HashSet

/**
 * Created on 18-3-28.
 *
 * @author leebin
 * @description Cookie 持久化「sp, db, file」
 */
class CookieJarImpl(context: Context?) : CookieJar {

  private val mPreferences: SharedPreferences

  override fun saveFromResponse(url: HttpUrl, cookies: List<Cookie>) {
    val values: MutableSet<String> = HashSet()
    for (cookie in cookies) {
      values.add(cookie.toString())
    }
    mPreferences.edit().putStringSet(url.host(), values).apply()
  }

  override fun loadForRequest(url: HttpUrl): List<Cookie> {
    val values = mPreferences.getStringSet(url.host(), HashSet())
    val cookies: MutableList<Cookie> = ArrayList()
    for (value in values!!) {
      cookies.add(Cookie.parse(url, value)!!)
    }
    return cookies
  }

  init {
    if (context == null) throw NullPointerException("context == null")
    mPreferences = context.getSharedPreferences("cookie-jar", Context.MODE_PRIVATE)
  }
}